
<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta charset="utf-8" />
    <title>Recapitulating the pPNA &#8212; pna_designer  documentation</title>
    <link rel="stylesheet" href="../_static/nature.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <script id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
    <script src="../_static/jquery.js"></script>
    <script src="../_static/underscore.js"></script>
    <script src="../_static/doctools.js"></script>
    <script src="../_static/language_data.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Managing Sequence Data from Silva" href="extra.html" />
    <link rel="prev" title="Recapitulating the mPNA" href="mPNA.html" /> 
  </head><body>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="extra.html" title="Managing Sequence Data from Silva"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="mPNA.html" title="Recapitulating the mPNA"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="../index.html">pna_designer  documentation</a> &#187;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="recapitulating-the-ppna">
<h1>Recapitulating the pPNA<a class="headerlink" href="#recapitulating-the-ppna" title="Permalink to this headline">¶</a></h1>
<p>General explanation of the procedure
MAPT, the Microbiome Amplification Preference Tool, is designed to find regions of homology between the host and its microbiome. MAPT has great flexibility, as any FASTA sequence can be used as the target sequence, i.e. the sequence that should be blocked. As you can see below, for the pPNA that sequence is A. thaliana, and is designed to recapitulate the results seen in Lundberg et al., 2013. MAPT users also specify the primers (shown below as the forward and reverse primer as Fprimer and Rprimer. This helps as only organisms that have homology to these primers are amplified in silico. The microbiome sequences come from Greengenes, as this was used to design the pPNA originally (Lundberg et al., 2013). K-mer range is the size that the sequences are fractionated into, such that K-mer range of 9-13 fractionates reads into sizes 9 to 13. By doing so we can identify regions that are ideal for targeting host amplification with oligonucleotides, such as PNAs, that will bind to host DNA and block amplification.</p>
<div class="section" id="required-information">
<h2>Required Information<a class="headerlink" href="#required-information" title="Permalink to this headline">¶</a></h2>
<p>Explanation of the materials needed (target sequence,
sequence file, primers, and k-mer range)</p>
<div class="highlight-ipython notranslate"><div class="highlight"><pre><span></span><span class="gp">In [1]: </span><span class="kn">import</span> <span class="nn">MAPT</span>

<span class="gp">In [2]: </span><span class="n">greengenes_fasta</span> <span class="o">=</span> <span class="s2">&quot;data/gg_97_otus_4feb2011.fasta&quot;</span>

<span class="gp">In [3]: </span><span class="n">Athaliana_chloro</span><span class="o">=</span><span class="s2">&quot;data/A_thaliana_chloroplast.fasta&quot;</span>

<span class="gp">In [4]: </span><span class="n">R926</span><span class="o">=</span><span class="s2">&quot;CCGYCAATTYMTTTRAGTTT&quot;</span>

<span class="gp">In [5]: </span><span class="n">F515</span> <span class="o">=</span> <span class="s2">&quot;GTGCCAGCMGCCGCGGTAA&quot;</span>

<span class="gp">In [6]: </span><span class="n">krng</span> <span class="o">=</span> <span class="p">(</span><span class="mi">9</span><span class="p">,</span><span class="mi">13</span><span class="p">)</span>

<span class="gp">In [7]: </span><span class="n">result_file</span><span class="o">=</span><span class="s1">&#39;data/pPNA.csv&#39;</span>
</pre></div>
</div>
<p>Now that we have all of our materials, we can begin mapping
k-mers from our greengenes data-set to the <em>A. thaliana</em> chloroplast</p>
</div>
<div class="section" id="generating-k-mer-maps">
<h2>Generating K-mer maps<a class="headerlink" href="#generating-k-mer-maps" title="Permalink to this headline">¶</a></h2>
<div class="highlight-ipython notranslate"><div class="highlight"><pre><span></span><span class="gp">In [8]: </span><span class="n">designer_pPNA</span> <span class="o">=</span> <span class="n">MAPT</span><span class="o">.</span><span class="n">PNA_Designer</span><span class="p">(</span><span class="n">result_file</span> <span class="o">=</span> <span class="s1">&#39;data/pPNA.csv&#39;</span><span class="p">,</span> <span class="n">target_fastafile</span> <span class="o">=</span> <span class="n">Athaliana_chloro</span><span class="p">,</span> <span class="n">sequence_file</span> <span class="o">=</span> <span class="n">greengenes_fasta</span><span class="p">,</span> <span class="n">primer_F</span><span class="o">=</span><span class="n">F515</span><span class="p">,</span><span class="n">primer_R</span> <span class="o">=</span> <span class="n">R926</span><span class="p">,</span> <span class="n">kmer_range</span> <span class="o">=</span> <span class="n">krng</span><span class="p">)</span>
<span class="go">Amplifying PNA target</span>
<span class="go">Amplifying and Collecting sequence K-mers</span>
<span class="go">	WARNING: 5150 sequences failed to amplify, please see attribute &quot;failed_amplification&quot;</span>
<span class="go">Mapping Kmers</span>
<span class="go">Mapping Complete</span>
<span class="go">Results have been written to data/pPNA.csv</span>
</pre></div>
</div>
<p>MAPT has automatically detected that some of the sequences
in our greengenes data-set failed to find <strong>exact</strong> primer matches. We
can view these sequences by looking at the ‘failed_amplification’ attribute
of MAPT.</p>
<p>With our k-mers mapped, we are able to visualize our data! But before
we do, we will quickly map the pPNA. While PNAs have a higher affinity
to the antiparallel direction, they can also bind in the parallel direction
as well. We will allow the PNA to bind in both directions for plotting</p>
</div>
<div class="section" id="mapping-a-pna">
<h2>Mapping a PNA<a class="headerlink" href="#mapping-a-pna" title="Permalink to this headline">¶</a></h2>
<div class="highlight-ipython notranslate"><div class="highlight"><pre><span></span><span class="gp">In [9]: </span><span class="n">pPNA</span> <span class="o">=</span> <span class="s2">&quot;GGCTCAACCCTGGACAG&quot;</span>

<span class="gp">In [10]: </span><span class="n">mapping</span> <span class="o">=</span> <span class="n">designer_pPNA</span><span class="o">.</span><span class="n">map_PNA</span><span class="p">(</span><span class="n">pPNA</span><span class="p">,</span><span class="n">antiparallel_only</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

<span class="gp">In [11]: </span><span class="kn">import</span> <span class="nn">pandas</span><span class="c1">#pandas is a module for working with dataframe objects in python, highly recommend!</span>

<span class="gp">In [12]: </span><span class="n">pandas</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">mapping</span><span class="p">)</span>
<span class="gh">Out[12]: </span><span class="go"></span>
<span class="go">    Nucleotide  index  unique match  absolute match  PNA mapping</span>
<span class="go">0            G      0            10          298066            0</span>
<span class="go">1            T      1            20          596146            0</span>
<span class="go">2            G      2            30          894270            0</span>
<span class="go">3            C      3            40         1192398            0</span>
<span class="go">4            C      4            50         1490502            0</span>
<span class="go">..         ...    ...           ...             ...          ...</span>
<span class="go">409          A    409            46         1067888            0</span>
<span class="go">410          C    410            36          805368            0</span>
<span class="go">411          G    411            26          542200            0</span>
<span class="go">412          G    412            16          274238            0</span>
<span class="go">413          G    413             6               8            0</span>

<span class="go">[414 rows x 5 columns]</span>
</pre></div>
</div>
<p>Now that we have data tables, use your favorite software to make
a figure! We will use a custom R script to plot our data.</p>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="../index.html">Table of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Recapitulating the pPNA</a><ul>
<li><a class="reference internal" href="#required-information">Required Information</a></li>
<li><a class="reference internal" href="#generating-k-mer-maps">Generating K-mer maps</a></li>
<li><a class="reference internal" href="#mapping-a-pna">Mapping a PNA</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="mPNA.html"
                        title="previous chapter">Recapitulating the mPNA</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="extra.html"
                        title="next chapter">Managing Sequence Data from Silva</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="../_sources/Pages/pPNA.rst.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" />
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="extra.html" title="Managing Sequence Data from Silva"
             >next</a> |</li>
        <li class="right" >
          <a href="mPNA.html" title="Recapitulating the mPNA"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="../index.html">pna_designer  documentation</a> &#187;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &#169; Copyright 2019, Spiridon Evangelos Papoulis.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 2.4.0.
    </div>
  </body>
</html>