
<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta charset="utf-8" />
    <title>Microbiome Amplification Preference Tool &#8212; pna_designer  documentation</title>
    <link rel="stylesheet" href="_static/nature.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="_static/language_data.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Recapitulating the gPNA" href="Pages/gPNA.html" /> 
  </head><body>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="Pages/gPNA.html" title="Recapitulating the gPNA"
             accesskey="N">next</a> |</li>
        <li class="nav-item nav-item-0"><a href="#">pna_designer  documentation</a> &#187;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="microbiome-amplification-preference-tool">
<h1>Microbiome Amplification Preference Tool<a class="headerlink" href="#microbiome-amplification-preference-tool" title="Permalink to this headline">¶</a></h1>
<p>Something general about the PNA</p>
<div class="section" id="check-how-we-recapitulated-the-design-of-different-pnas">
<h2>Check how we recapitulated the design of different PNAs!<a class="headerlink" href="#check-how-we-recapitulated-the-design-of-different-pnas" title="Permalink to this headline">¶</a></h2>
<div class="toctree-wrapper compound">
<ul>
<li class="toctree-l1"><a class="reference internal" href="Pages/gPNA.html">Recapitulating the gPNA</a></li>
<li class="toctree-l1"><a class="reference internal" href="Pages/mPNA.html">Recapitulating the mPNA</a></li>
<li class="toctree-l1"><a class="reference internal" href="Pages/pPNA.html">recapitulating the pPNA</a></li>
</ul>
</div>
</div>
<div class="section" id="design-a-pna">
<h2>Design a PNA<a class="headerlink" href="#design-a-pna" title="Permalink to this headline">¶</a></h2>
<p>PNA_Designer is the primary class for desiging
your own custom PNA.</p>
<dl class="class">
<dt id="MAPT.PNA_Designer">
<em class="property">class </em><code class="sig-prename descclassname">MAPT.</code><code class="sig-name descname">PNA_Designer</code><span class="sig-paren">(</span><em class="sig-param">result_file=''</em>, <em class="sig-param">target_silva_accession=''</em>, <em class="sig-param">target_fastafile=''</em>, <em class="sig-param">sequence_file=''</em>, <em class="sig-param">sequence_silva_path=0</em>, <em class="sig-param">primer_F=''</em>, <em class="sig-param">primer_R=''</em>, <em class="sig-param">kmer_range=(9</em>, <em class="sig-param">14)</em>, <em class="sig-param">silva_release=0</em>, <em class="sig-param">silva_dataset='ref'</em>, <em class="sig-param">subunit='small'</em>, <em class="sig-param">database_dir=''</em><span class="sig-paren">)</span><a class="headerlink" href="#MAPT.PNA_Designer" title="Permalink to this definition">¶</a></dt>
<dd><p>A tool used to streamline the design of PNA oligos used for blocking
amplificaiton of specific DNAs during PCR amplificaiton</p>
<p>PNA_designer is a module that automates key steps in designing a PNA blocker. Unfortunatly,
no single PNA is universal in blocking DNA, thus, each requires testing in order to determine
the sequence space being blocked during amplificaiton. This tool automates inital <em>in silico</em>
analysis required for development of a new PNA. To complement automatic analysis, this tool
also makes data access more convient by utilizing RNA sequence data avalible
at the <a class="reference external" href="https://www.arb-silva.de/">Silva database</a>. Silva identifiers can be used to
specify either the PNA target sequence or sequences concidered in the analysis. A copy of silva
datasets will be download and then compiled.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>results_file</strong> (<em>str</em>) – The name of the output file</p></li>
<li><p><strong>target_silva_accession</strong> (<em>str, optional</em>) – Silva accession of PNA target. Required if “target_fastafile” is not provided.</p></li>
<li><p><strong>target_fastafile</strong> (<em>str, optional</em>) – Fasta file with PNA target. File should only have PNA target sequence. Required if
“target_silva_accession” is not provided.</p></li>
<li><p><strong>sequence_file</strong> (<em>str, optional</em>) – Fasta file used to desgin the PNA. Required if “sequence_silva_taxid” not provided.</p></li>
<li><p><strong>sequence_silva_taxid</strong> (<em>int, optional</em>) – Silva taxid used to desgin the PNA. Required if “sequence_silva_taxid” not provided.</p></li>
<li><p><strong>primer_F</strong> (<em>str, optional</em>) – Forward primer used for the in silico amplification. If not specified, the full sequence
of the PNA target and sequences for PNA design will be used.</p></li>
<li><p><strong>primer_R</strong> (<em>str, optional</em>) – Reverse primer used for the in silico amplification. If not specified, the full sequence
of the PNA target and sequences for PNA design will be used.</p></li>
<li><p><strong>kmer_range</strong> (<em>tupple, optional</em>) – Kmer range used in PNA design. Defaults range: 9-14</p></li>
<li><p><strong>silva_release</strong> (<em>int, optional</em>) – Specify Silva release to be used. Only release &gt;=132 avalible. Default = Current Release.</p></li>
<li><p><strong>silva_dataset</strong> (<em>str, optional</em>) – Silva Database used to fetch seqeucnes. Avalible options are <strong>parc</strong> (all sequences),
<strong>ref</strong> (high quality sequences), or <strong>nr</strong> (non-redundant, clustered
99% identity criterion). Defaults to <strong>nr</strong></p></li>
<li><p><strong>subunit</strong> (<em>str, optional</em>) – Specify <strong>large</strong> or <strong>small</strong> ribosomal subunit in silva database. Defaults to <strong>small</strong></p></li>
<li><p><strong>database_dir</strong> (<em>str, optional</em>) – Local directory of silva database if “target_silva_accession” or
“sequence_silva_taxid” is specified. Defualt locaiton is “~/.pna_designer”</p></li>
</ul>
</dd>
<dt class="field-even">Variables</dt>
<dd class="field-even"><p><strong>failed_amplification</strong> (<em>list of str</em>) – If primers are specified, accession’s of sequences that do not have an exact match
to the primers will be recorded</p>
</dd>
</dl>
<dl class="method">
<dt id="MAPT.PNA_Designer.map_PNA">
<code class="sig-name descname">map_PNA</code><span class="sig-paren">(</span><em class="sig-param">PNA</em>, <em class="sig-param">antiparallel_only=False</em><span class="sig-paren">)</span><a class="headerlink" href="#MAPT.PNA_Designer.map_PNA" title="Permalink to this definition">¶</a></dt>
<dd><p>Align a PNA to a given target sequence</p>
<p>map_PNA will use the k_mapper class to generate k-mer alignments to target DNA. PNA oligomers
can bind to DNA in either orientation, thus, the reverse of the PNA is also searched for alignments</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>PNA</strong> (<em>str</em>) – PNA sequence</p></li>
<li><p><strong>antiparallel_only</strong> (<em>bool, optional</em>) – Specifies if only the antiparallel orientation should be concidered in alignment. Defaults to False</p></li>
</ul>
</dd>
</dl>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="additional-resources-in-this-module">
<h2>Additional Resources in this module<a class="headerlink" href="#additional-resources-in-this-module" title="Permalink to this headline">¶</a></h2>
<p>To increase the tractibility offered to researchers, wes
have made the k_mapper datatype avalible for users.</p>
<div class="toctree-wrapper compound">
<ul>
<li class="toctree-l1"><a class="reference internal" href="Pages/extra.html">Managing Sequence Data from Silva</a><ul class="simple">
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="Pages/extra.html#mapping-kmers">Mapping Kmers</a><ul class="simple">
</ul>
</li>
</ul>
</div>
</div>
</div>
<div class="section" id="indices-and-tables">
<h1>Indices and tables<a class="headerlink" href="#indices-and-tables" title="Permalink to this headline">¶</a></h1>
<ul class="simple">
<li><p><a class="reference internal" href="genindex.html"><span class="std std-ref">Index</span></a></p></li>
<li><p><a class="reference internal" href="py-modindex.html"><span class="std std-ref">Module Index</span></a></p></li>
<li><p><a class="reference internal" href="search.html"><span class="std std-ref">Search Page</span></a></p></li>
</ul>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="#">Table of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Microbiome Amplification Preference Tool</a><ul>
<li><a class="reference internal" href="#check-how-we-recapitulated-the-design-of-different-pnas">Check how we recapitulated the design of different PNAs!</a></li>
<li><a class="reference internal" href="#design-a-pna">Design a PNA</a></li>
<li><a class="reference internal" href="#additional-resources-in-this-module">Additional Resources in this module</a></li>
</ul>
</li>
<li><a class="reference internal" href="#indices-and-tables">Indices and tables</a></li>
</ul>

  <h4>Next topic</h4>
  <p class="topless"><a href="Pages/gPNA.html"
                        title="next chapter">Recapitulating the gPNA</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/index.rst.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" />
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="Pages/gPNA.html" title="Recapitulating the gPNA"
             >next</a> |</li>
        <li class="nav-item nav-item-0"><a href="#">pna_designer  documentation</a> &#187;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &#169; Copyright 2019, Spiridon Evangelos Papoulis.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 2.2.0.
    </div>
  </body>
</html>